plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePageSink.java

```diff
@@ -251,29 +243,38 @@ public void abort()
     {
         // Must be wrapped in doAs entirely
         // Implicit FileSystem initializations are possible in HiveRecordWriter#rollback -> RecordWriter#close
-        hdfsEnvironment.doAs(session.getIdentity(), this::doAbort);
+        hdfsEnvironment.doAs(session.getIdentity(), (Runnable) this::doAbort);
     }
 
     private void doAbort()
     {
         Optional<Exception> rollbackException = Optional.empty();
-        for (HiveWriter writer : Iterables.concat(writers, closedWriters)) {
+        for (HiveWriter writer : writers) {
             // writers can contain nulls if an exception is thrown when doAppend expends the writer list
             if (writer != null) {
-                try {
-                    writer.rollback();
-                }
-                catch (Exception e) {
-                    log.warn("exception '%s' while rollback on %s", e, writer);
-                    rollbackException = Optional.of(e);
-                }
+                rollbackException = doAbort(new RollbackAction(writer.getFileWriter()), rollbackException);
             }
         }
+        for (RollbackAction rollbackAction : closedWriterRollbackActions) {
+            rollbackException = doAbort(rollbackAction, rollbackException);
+        }
         if (rollbackException.isPresent()) {
             throw new TrinoException(HIVE_WRITER_CLOSE_ERROR, "Error rolling back write to Hive", rollbackException.get());
         }
     }
 
+    private Optional<Exception> doAbort(RollbackAction rollbackAction, Optional<Exception> rollbackException)
+    {
+        try {
+            rollbackAction.rollback();
+        }
+        catch (Exception e) {
+            log.warn("exception '%s' while rollback on %s", e, rollbackAction);
```

Comment:
The original behavior of this code is not good, so we should fix it while we refactor. This is both logging the exception and then throwing it. Also, it arbitrarily throws the last exception. We should either

1. Log the exception and discard it 
2. Build a single exception with all of the exceptions attached as suppressions and don't log